File: database-driver-plugin/src/main/java/de/lino/database/utility/cache/DefaultClusteredCache.java
package de.lino.database.utility.cache;

import de.lino.database.utils.cache.Cache;
import de.lino.database.utils.cache.ClusteredCache;
import de.lino.database.utils.cache.ConsistentHashRing;

import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Function;

/**
 * Default implementation of {@link ClusteredCache}, partitioning entities
 * across multiple {@link DefaultCache} shards using {@link ConsistentHashRing}
 * for routing and replication, in pure Java with no external dependencies
 * beyond {@code java.*}.
 * <p>
 * <b>Honest note on scale:</b> this class simulates Cassandra's partitioning
 * principle WITHIN a single JVM. For real 10<sup>12</sup>-scale distribution across
 * multiple physical machines, network communication between JVM processes would
 * additionally be required (e.g. via {@code java.net.Socket} or
 * {@code java.net.http.HttpClient}, both already part of the JDK with no further
 * dependency) — each shard would then run as its own process on its own hardware.
 * The routing principle (this ring) stays identical; only access to a given node
 * would go over the network instead of a direct method call.
 *
 * @param <ID> key type
 * @param <T>  entity type
 */
public final class DefaultClusteredCache<ID, T> implements ClusteredCache<ID, T> {

    private final ConsistentHashRing<Integer> ring;
    private final ConcurrentHashMap<Integer, Cache<ID, T>> shards = new ConcurrentHashMap<>();
    private final int replicationFactor;

    /**
     * @param shardCount        number of simulated nodes (e.g. CPU core count, or —
     *                          in true distribution — machine count)
     * @param replicationFactor number of shards each key is simultaneously stored on
     * @param loader            supplies an entity if it is not cached on any shard
     * @param ttl               validity duration per entry
     * @param maxSizePerShard   size limit PER shard (total capacity = shardCount * maxSizePerShard)
     */
    public DefaultClusteredCache(int shardCount, int replicationFactor,
                                  Function<ID, CompletableFuture<T>> loader,
                                  Duration ttl, long maxSizePerShard) {
        if (shardCount < 1) throw new IllegalArgumentException("shardCount must be >= 1");
        if (replicationFactor < 1 || replicationFactor > shardCount) {
            throw new IllegalArgumentException("replicationFactor must be between 1 and shardCount");
        }
        Objects.requireNonNull(loader, "loader must not be null");

        this.replicationFactor = replicationFactor;

        List<Integer> nodeIds = new ArrayList<>();
        for (int i = 0; i < shardCount; i++) {
            nodeIds.add(i);
            shards.put(i, new DefaultCache<>(loader, ttl, maxSizePerShard));
        }
        this.ring = new DefaultConsistentHashRing<>(nodeIds);
    }

    @Override
    public CompletableFuture<T> get(ID id) {
        Objects.requireNonNull(id, "id must not be null");
        int primaryShard = ring.nodeFor(id);
        return shards.get(primaryShard).get(id);
    }

    @Override
    public CompletableFuture<Void> put(ID id, T value) {
        Objects.requireNonNull(id, "id must not be null");
        Objects.requireNonNull(value, "value must not be null");

        List<Integer> targetShards = ring.nodesFor(id, replicationFactor);
        CompletableFuture<?>[] writes = targetShards.stream()
                .map(shardId -> CompletableFuture.runAsync(() -> shards.get(shardId).put(id, value)))
                .toArray(CompletableFuture[]::new);
        return CompletableFuture.allOf(writes);
    }

    @Override
    public void invalidate(ID id) {
        Objects.requireNonNull(id, "id must not be null");
        for (int shardId : ring.nodesFor(id, replicationFactor)) {
            shards.get(shardId).invalidate(id);
        }
    }

    @Override
    public long totalSize() {
        return shards.values().stream().mapToLong(Cache::size).sum();
    }

    @Override
    public int shardCount() {
        return shards.size();
    }
}

File: python/database-driver-plugin/src/database_driver/plugin/utility/cache/default_clustered_cache.py
"""Mirror of ``de.lino.database.utility.cache.DefaultClusteredCache``."""

from __future__ import annotations

import asyncio
from collections.abc import Awaitable, Callable
from datetime import timedelta
from typing import TypeVar

from database_driver.api.utils.cache.cache import Cache
from database_driver.api.utils.cache.clustered_cache import ClusteredCache
from database_driver.plugin.utility.cache.default_cache import DefaultCache
from database_driver.plugin.utility.cache.default_consistent_hash_ring import DefaultConsistentHashRing

ID = TypeVar("ID")
T = TypeVar("T")


class DefaultClusteredCache(ClusteredCache[ID, T]):
    """Default implementation of ``ClusteredCache``, partitioning entities across
    multiple ``DefaultCache`` shards using ``DefaultConsistentHashRing`` for routing and
    replication, in pure Python with no dependencies beyond the stdlib.

    **Honest note on scale:** this class simulates Cassandra's partitioning principle
    WITHIN a single process. For real distribution across multiple physical machines,
    network communication between processes would additionally be required - each shard
    would then run as its own process on its own hardware. The routing principle (this
    ring) stays identical; only access to a given node would go over the network instead
    of a direct method call.
    """

    def __init__(
        self,
        shard_count: int,
        replication_factor: int,
        loader: Callable[[ID], Awaitable[T]],
        ttl: timedelta | None,
        max_size_per_shard: int,
    ) -> None:
        """Args:
            shard_count: Number of simulated nodes (e.g. CPU core count, or - in true
                distribution - machine count).
            replication_factor: Number of shards each key is simultaneously stored on.
            loader: Supplies an entity if it is not cached on any shard.
            ttl: Validity duration per entry.
            max_size_per_shard: Size limit PER shard (total capacity =
                shard_count * max_size_per_shard).
        """
        if shard_count < 1:
            raise ValueError("shard_count must be >= 1")
        if replication_factor < 1 or replication_factor > shard_count:
            raise ValueError("replication_factor must be between 1 and shard_count")
        if loader is None:
            raise TypeError("loader must not be None")

        self._replication_factor = replication_factor
        self._shards: dict[int, Cache[ID, T]] = {
            index: DefaultCache(loader, ttl, max_size_per_shard) for index in range(shard_count)
        }
        self._ring = DefaultConsistentHashRing(list(range(shard_count)))

    async def get(self, id: ID) -> T:
        if id is None:
            raise TypeError("id must not be None")
        primary_shard = self._ring.node_for(id)
        return await self._shards[primary_shard].get(id)

    async def put(self, id: ID, value: T) -> None:
        """Writes to all replica shards responsible for this key concurrently, not
        sequentially - the analogue of the Java edition fanning the per-shard writes out
        on the common pool."""
        if id is None:
            raise TypeError("id must not be None")
        if value is None:
            raise TypeError("value must not be None")

        target_shards = self._ring.nodes_for(id, self._replication_factor)
        await asyncio.gather(
            *(asyncio.to_thread(self._shards[shard_id].put, id, value) for shard_id in target_shards)
        )

    def invalidate(self, id: ID) -> None:
        if id is None:
            raise TypeError("id must not be None")
        for shard_id in self._ring.nodes_for(id, self._replication_factor):
            self._shards[shard_id].invalidate(id)

    def total_size(self) -> int:
        return sum(shard.size() for shard in self._shards.values())

    def shard_count(self) -> int:
        return len(self._shards)
